packages/template-retail-react-app/app/hooks/use-datacloud.js

```diff
@@ -0,0 +1,481 @@
+        }
+    }
+}
+
+/**
+ * Custom hook for sending PWA Kit events to Data Cloud.
+ *
+ * This hook provides methods to track various user interactions, such as
+ * page views, product views, category views, search impressions, and recommendations.
+ *
+ * @returns {object} An object containing methods for sending different Data Cloud events
+ */
+const useDataCloud = () => {
+    const {getUsidWhenReady} = useUsid()
+    const {isRegistered} = useCustomerType()
+    const customerId = useCustomerId() // Bug PWA -> Needs converted to Promise
+    const {data: customer} = useCurrentCustomer()
+    const {site} = useMultiSite()
+    let {effectiveDnt} = useDNT()
+    effectiveDnt = false // Remove after demo
+
+    const getEventUserParameters = async () => {
+        return {
+            isGuest: isRegistered ? 0 : 1,
+            usid: await getUsidWhenReady(),
+            customerId: customerId,
+            customerNo: customer?.customerNo,
+            firstName: customer?.firstName,
+            lastName: customer?.lastName,
+            email: customer?.email
+        }
+    }
+
+    const {
+        app: {dataCloudAPI: config, defaultSite: siteId}
+    } = getConfig()
+
+    const {appSourceId, tenantId} = config
+    let sdk = null
+    let isDatacloudInitiated = true
+
+    if (!appSourceId || !tenantId) {
+        isDatacloudInitiated = false
+    } else {
+        sdk = initDataCloudSdk(tenantId, appSourceId)
+    }
+
+    const dataCloud = useMemo(
+        () =>
+            new DataCloudApi({
+                site: site.id,
+                sdk
+            }),
+        [site, sdk]
+    )
+
+    return {
+        async sendViewPage(...args) {
+            if (!isDatacloudInitiated || effectiveDnt) return
+            const userParameters = await getEventUserParameters()
+            return dataCloud.sendViewPage(...args.concat(userParameters))
+        },
+        async sendViewProduct(...args) {
+            if (!isDatacloudInitiated || effectiveDnt) return
+            const userParameters = await getEventUserParameters()
+            return dataCloud.sendViewProduct(...args.concat(userParameters))
+        },
+        async sendViewCategory(...args) {
+            if (!isDatacloudInitiated || effectiveDnt) return
+            const userParameters = await getEventUserParameters()
+            return dataCloud.sendViewCategory(...args.concat(userParameters))
+        },
+        async sendViewSearchResults(...args) {
+            if (!isDatacloudInitiated || effectiveDnt) return
+            const userParameters = await getEventUserParameters()
+            return dataCloud.sendViewSearchResults(...args.concat(userParameters))
+        },
+        async sendViewRecommendations(...args) {
+            if (!isDatacloudInitiated || effectiveDnt) return
+            const userParameters = await getEventUserParameters()
+            return dataCloud.sendViewRecommendations(...args.concat(userParameters))
```

Comment:
Can we create an variable for `...args.concat(userParameters)` instead of having to repeat it here?